### PR TITLE
Remove the big q splitting rule

### LINKED ISSUE #617: Ensure small q / big Q splitting rule results are identical
Currently, for performance reasons, trees call into different splitting functions (small q or big Q) depending on the number of samples in the candidate node. 

There are no explicit tests that ensure these two functions give identical results. Implicit checks can be made to be sure they are identical:

1. In ForestCharacterizationTests.cpp enable UPDATE_PREDICTION_FILES
2. In globals.h set Q_THRESHOLD to a large number, e.g. 9999999999 to only activate the small q splitting rule
3. Compile and run the executable to update all test output files
4. In ForestCharacterizationTests.cpp disable UPDATE_PREDICTION_FILES
5. In globals.h set Q_THRESHOLD to a large negative number, e.g. -9999999999 to only activate the big Q splitting rule
6. Compile and run the executable, if all characterization tests pass the two splitting functions produce identical output for the forest test cases listed in ForestCharacterizationTests.cpp

This check passes with the latest major splitting rule modification #590. One caveat is that this equality holds almost surely, subject to the confines of floating point precision, as mentioned in https://github.com/grf-labs/grf/pull/590#issuecomment-584339016.

Future test coverage could be extended by explicit tests of this behavior.

### PR DESCRIPTION
**Pull request by erikcs**
As discussed earlier, remove the "large q" splitting rule. With #592 (small q improvement) there is not a big performance benefit (considering the added complexity) to keeping both.

Also address some technical debt, add a pseudocode section at the end of [DEVELOPING.md](https://github.com/erikcs/grf/blob/remove-bigQ/DEVELOPING.md#details) which explains the current regression splitting with pseudocode. 

_Details_:

* `Data.sort()` and all its components are removed

* ~Also remove `possible_split_values`, follow up from https://github.com/grf-labs/grf/pull/592#discussion_r360650548~

_Checks_:

* [causal_benchmark.txt](https://gist.github.com/erikcs/202a80081aa04d9ba7df108eaee85dbd)

(this closes #617 which is no longer necessary) 

_Future follow ups_:

There are future room for improvements concerning a) code clarity and design concerning splitting rules - components that may be easier to unit test b) low-level optimizations, for example, STL vector containers (C++11) do not take advantage of SIMD instructions (but some does in C++17), Eigen does in some cases. 

(Also implement the scratch array size reduction as suggested below https://github.com/grf-labs/grf/pull/635#discussion_r395917188)

### CODE REVIEW

File: core/src/commons/Data.cpp [hunk 1]
```diff
@@ -160,11 +158,11 @@ void Data::set_weight_index(size_t index) {
   disallowed_split_variables.insert(index);
 }
 
-void Data::get_all_values(std::vector<double>& all_values,
+void Data::get_all_values(size_t& num_unique_samples,
```
Comment (halflearned) [inline core/src/commons/Data.cpp]: I've never seen a `size_t` being passed as a _reference_, in particular as non-const. I imagine the reason for this choice was that it allows us  keep this function's return type as `void`. Could we instead, e.g., return a struct with all the information we want?

File: core/src/splitting/InstrumentalSplittingRule.cpp [hunk 2]
```diff
@@ -127,33 +119,32 @@ bool InstrumentalSplittingRule::find_best_split(const Data& data,
...
-                                                              double sum_node_z_squared,
-                                                              double min_child_size,
-                                                              double& best_value,
-                                                              size_t& best_var,
-                                                              double& best_decrease,
-                                                              bool& best_send_missing_left,
-                                                              const std::vector<double>& responses_by_sample,
-                                                              const std::vector<std::vector<size_t>>& samples) {
-  std::vector<double> possible_split_values;
+void InstrumentalSplittingRule::find_best_split_value(const Data& data,
```
Comment (halflearned) [inline core/src/splitting/InstrumentalSplittingRule.cpp]: [nice](https://i.imgflip.com/1bk7m4.jpg)

File: core/src/splitting/InstrumentalSplittingRule.cpp [hunk 3]
```diff
@@ -285,7 +276,8 @@ void InstrumentalSplittingRule::find_best_split_value_small_q(const Data& data,
 
       // Save this split if it is the best seen so fa.
```
Comment (halflearned) [inline core/src/splitting/InstrumentalSplittingRule.cpp]:
```suggestion
      // Save this split if it is the best seen so far.
```

File: DEVELOPING.md [hunk 4]
```diff
@@ -53,3 +53,78 @@ When designing a forest for a new statistical task, we suggest following the for
...
+weighted_response_sums_left += weighted_response_sum_missing
+
+weight_sums_right = total_weight_sum - weight_sums_left
+sample_counters_right = total_samples - sample_counters_left
+weighted_response_sums_right = total_sum - weighted_response_sums_left
+
+decrease_na_left = weighted_response_sums_left^2 / weight_sums_left +
+                   weighted_response_sums_right^2 / weight_sums_right
+
+if max(decrease_na_left) > max(decrease_na_right):
```
[the remark this answers is not in the document]
Comment (davidahirshberg) [reply DEVELOPING.md]: This pseudocode was hard for me to understand. The detailed computations obscure the meaning of what is happening. I think something like the following would be easier on readers:

for each unique value u in X
   let  count_left[u] be the number of samples with X <= u 
   let  sum_left[u], weight_sum_left[u] be the sum of weight*response and weight over those elements
   and let *_right[u], *_missing be the same for samples with X > u and X missing.
   
   decrease[u, missing=on-right] = sum_left[u]^2 / weight_sum_left[u] + sum_right[u]^2 / weight_sums_right[u]
   decrease[u, missing=on-left] = ...

return (u, missing) that maximizing decrease among u with count_left[u] and count_right[u] >= min-child-size

It may also help to have an explanation for why decrease has the form it does, with a square in the numerator and none in the denominator.

File: core/src/commons/Data.cpp (hunk 1 above)
Comment (davidahirshberg) [reply core/src/commons/Data.cpp]: Looks like idiomatic C++ to me. Maybe I'm stuck in the early 00s.

Comment (erikcs) [reply core/src/commons/Data.cpp]: Yes, it's to keep it consistent with the rest of grf, like `find_best_split_value`

File: DEVELOPING.md [hunk 5]
```diff
@@ -53,3 +53,78 @@ When designing a forest for a new statistical task, we suggest following the for
 To avoid the overhead of setting up new classes and Rcpp bindings, we provide a template for a 'custom' forest. To get started, fill in the implementation for CustomRelabelingStrategy and CustomPredictionStrategy. By default, the custom forest uses the standard regression splitting rule -- if you'd like to change this, consult the ForestTrainers.custom_trainer method.
 
 This forest is made available in R as `custom.forest` (with `predict.custom.forest`). You can find a starter template for a test exercising the forest in `testthat/test_custom_forest.R`. Note that you'll need to re-run `build_package.R` after making changes to the C++ source.
+
```
[the remark this answers is not in the document]
Comment (erikcs) [reply DEVELOPING.md]: I thought this could more serve as code documentation, and Imke will add a paragraph on MIA to REFERENCE, which is more big picture algorithmic details?

File: DEVELOPING.md (hunk 4 above)
[the remark this answers is not in the document]
Comment (erikcs) [reply DEVELOPING.md]: Thanks, it's been simplified. Do you have a suggestion for a decrease explanation?

File: core/src/commons/Data.cpp [hunk 6]
```diff
@@ -176,66 +174,32 @@ void Data::get_all_values(std::vector<double>& all_values,
   std::iota(index.begin(), index.end(), 0);
   // sort index based on the split values (argsort)
   // the NaN comparison places all NaNs at the beginning
+  // stable sort is needed for consistent element ordering cross platform,
+  // otherwise the resulting sums used in the splitting rules may compound rounding error
+  // differently and produce different splits.
   std::stable_sort(index.begin(), index.end(), [&](const size_t& lhs, const size_t& rhs) {
-    return all_values[lhs] < all_values[rhs] || (std::isnan(all_values[lhs]) && !std::isnan(all_values[rhs]));
+    double left_val = all_values[lhs];
```
Comment (jtibshirani) [inline core/src/commons/Data.cpp]: What's the motivation for the change here, and for the related changes in the splitting rules? Is it just a straight refactor? If so it'd be best to leave it out of this PR, since it doesn't seem related to removing the big Q splitting rule. (To me the refactor isn't an obvious improvement either, as it adds complexity into the splitting rules around retrieving the right 'best split value').

File: DEVELOPING.md [hunk 7]
```diff
@@ -53,3 +53,41 @@ When designing a forest for a new statistical task, we suggest following the for
...
+### Details
```
Comment (jtibshirani) [inline DEVELOPING.md]: This is really helpful! For simplicity, I think we could just call this section 'Tree Splitting Algorithm' instead of 'Details' and remove the 'SplittingRule' header below.

File: core/src/commons/Data.cpp (hunk 6 above)
Comment (erikcs) [reply core/src/commons/Data.cpp]: This simple change was just meant to remove two additional (potentially) cache expensive retrievals of all_values[.], that was probably unnecessary the compiler should optimize that away? I can restore it to the original.

The change in the splitting rule comes from removing the use of `possible_split_values` ( skipping an `erase` call). You mentioned it here https://github.com/grf-labs/grf/pull/592#discussion_r360650548 and I thought it made sense to do here, should I restore it? (I can just revert commit b4ab0af)

File: DEVELOPING.md (hunk 5 above)
[the remark this answers is not in the document]
Comment (erikcs) [reply DEVELOPING.md]: (Marking this as resolved as Julie seemed OK with having it in DEVELOPING)

File: core/src/splitting/factory/InstrumentalSplittingRuleFactory.cpp [hunk 8]
```diff
@@ -23,7 +23,7 @@ namespace grf {
 std::unique_ptr<SplittingRule> InstrumentalSplittingRuleFactory::create(const Data& data,
                                                                         const TreeOptions& options) const {
   return std::unique_ptr<SplittingRule>(new InstrumentalSplittingRule(
-      data.get_max_num_unique_values(),
+      data.get_num_rows(),
```
[the remark this answers is not in the document]
Comment (erikcs) [reply core/src/splitting/factory/InstrumentalSplittingRuleFactory.cpp]: That sounds very worthwhile, do you have design thoughts how to put that in place?

Right before the splitting rule factory is called we know the sub sample sample size (the length of` nodes[0]` in the previous line):

https://github.com/grf-labs/grf/blob/master/core/src/tree/TreeTrainer.cpp#L60

We can move the scratch arrays initialization to the Factory? (I can do that in a follow up PR)

File: core/src/commons/Data.cpp (hunk 6 above)
Comment (jtibshirani) [reply core/src/commons/Data.cpp]: My original comment was unclear -- by 'the change here' I meant the changes to the method as a whole. I think it'd make sense to revert the changes to the method (including commit b4ab0af as you mention), since it's separate from the goal of the PR. And I'm not sure it helps simplify or clarify the logic now that we have to make changes to the splitting rules.

File: core/src/splitting/factory/InstrumentalSplittingRuleFactory.cpp (hunk 8 above)
[the remark this answers is not in the document]
Comment (jtibshirani) [reply core/src/splitting/factory/InstrumentalSplittingRuleFactory.cpp]: Could we change the signature of `SplittingRuleFactory` to take in the subsample size instead of `Data`? And sounds good to discuss and make the change in a follow-up PR.

File: core/src/commons/Data.cpp (hunk 6 above)
Comment (erikcs) [reply core/src/commons/Data.cpp]: Ok, it has been reverted

File: core/src/splitting/factory/InstrumentalSplittingRuleFactory.cpp (hunk 8 above)
[the remark this answers is not in the document]
Comment (erikcs) [reply core/src/splitting/factory/InstrumentalSplittingRuleFactory.cpp]: 👍

File: DEVELOPING.md (hunk 4 above)
[the remark this answers is not in the document]
Comment (erikcs) [reply DEVELOPING.md]: I'll add a decrease explanation here over the weekend, thanks @halflearned for the algebra clarification.

[the remark this answers is not in the document]
Comment (erikcs) [reply DEVELOPING.md]: A decrease explanation is now added to DEVELOPING, if this is OK I'll go ahead and merge.
